rack-protection/lib/rack/protection/cookie_tossing.rb

```diff
@@ -31,8 +31,7 @@ def accepts?(env)
         cookie_header = env['HTTP_COOKIE']
         cookies = Rack::Utils.parse_query(cookie_header, ';,') { |s| s }
         cookies.each do |k, v|
-          if (k == session_key && Array(v).size > 1) ||
-             (k != session_key && Rack::Utils.unescape(k) == session_key)
+          if k == session_key && Array(v).size > 1
```

Comment:
I think the point of this middleware is to deny requests with both `rack.session` and `rack.%73ession` cookie keys. If we don't unescape, we fail to detect that. Right?

Reply:
Those cookie keys don't collide in Rack either, as we don't unescape cookie keys so they are distinct keys.

Reply:
In Rack <3.1 they do collide?

Reply:
It looks okay to me, IOW, it can't collide.

- 3-1-stable: https://github.com/rack/rack/blob/3-1-stable/lib/rack/utils.rb#L238-L239
- 3-0-stable: https://github.com/rack/rack/blob/3-0-stable/lib/rack/utils.rb#L250-L251
- 2-2-stable: https://github.com/rack/rack/blob/2-2-stable/lib/rack/utils.rb#L232-L233
- 2-1-stable: https://github.com/rack/rack/blob/2-1-stable/lib/rack/utils.rb#L232-L233

This was a security bug, and was fixed in various stable branches, e.g.

- 2-1-stable fixed in https://github.com/rack/rack/commit/a2435106dc1216d8785fcac68ef53030bb484b92
- 2-2-stable fixed in https://github.com/rack/rack/commit/5ccca4722668083732ea2d35c56565fcc25312f8
- 3-0-stable: https://github.com/rack/rack/commit/1f5763de6a9fe515ff84992b343d63c88104654c
- 3-1-stable: same as above.

I understand this is an important issue, so feel free to do your own investigation and add appropriate tests to confirm the behaviour.

Reply:
So what this change does it having the CookieTossing middleware accept more requests, but it should be safe to do so.

Reply:
Yes, the previous key collisions can no longer occur.